File: product of array except itself.py
# TC and SC: O(n)

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n=len(nums)
        
        leftArray=[0]*n
        rightArray=[0]*n
        
        ans=[0]*n
        
        leftArray[0]=1
        
        for i in range(1,n):
            leftArray[i]=leftArray[i-1]*nums[i-1]
        
        
        rightArray[n-1]=1
        for i in reversed(range(n-1)):
            rightArray[i]=rightArray[i+1]*nums[i+1]
        
        #print(leftArray,rightArray)
        
        for i in range(n):
            ans[i]=leftArray[i]*rightArray[i]
            
        return ans

File: Product of Array except itself.java
// Time Complexity - O(2n) = O(n) 
// Space Complexity - O(1) since no auxillary space has been used, only using output

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int [] results = new int [n];
        int rp = 1;
        results[0] = 1;
        // left repeating prouduct
        for(int i =1 ; i<= n-1; i++){
            rp = rp * nums[i-1];
            results[i] = rp;
        }
        rp = 1;
        // right repeating product
        for(int j = n-2; j>=0; j--){
            rp = rp * nums[j+1];
            results[j] = results[j] * rp;
        }
        
        return results;
        
    }
}